src/web/src/app/space/add/component.ts

```diff
@@ -0,0 +1,135 @@
+import { startWith, map, debounceTime, distinctUntilChanged, switchMap } from 'rxjs/operators';
+import { MatOption } from '@angular/material/core';
+import { Router, NavigationExtras } from '@angular/router';
+import { FormGroup, FormBuilder, Validators } from '@angular/forms';
+
+import { SpaceService } from '../service';
+import { Space } from '../model';
+import { SpaceKeyValidation } from './space-key-validator';
+import { RequestWatcherService } from '../../request-watcher/service';
+
+@Component({
+	selector: 'app-space-add',
+	templateUrl: './component.html',
+	styleUrls: ['./component.css'],
+})
+export class SpaceAddComponent implements OnInit {
+	createForm: FormGroup;
+	controlAutocomplete = new FormControl('');
+
+	spaces: Observable<Space[]>;
+	selectedSpace: Space;
+	keyWasChanged: boolean = false;
+
+	constructor(
+		private router: Router,
+		private fb: FormBuilder,
+		private space: SpaceService,
+		private spaceKeyValidator: SpaceKeyValidation,
+		private requestWatcher: RequestWatcherService
+	) {
+		this._createForm();
+	}
+
+	private translitFromRuToEn = {
+	"ё": "yo", "й": "i", "ц": "ts", "у": "u", "к": "k", "е": "e", "н": "n", "г": "g",
+	"ш": "sh", "щ": "sch", "з": "z", "х": "h", "ъ": "'", "ф": "f", "ы": "i", "в": "v", "а": "a", "п": "p", "р": "r", "о": "o", "л": "l", "д": "d", "ж": "zh", "э": "e", "я": "ya", "ч": "ch", "с": "s", "м": "m", "и": "i", "т": "t", "ь": "'", "б": "b",
+	"ю": "yu"
+	};
+
+	ngOnInit() {
+
+		this.spaces = this.controlAutocomplete.valueChanges.pipe(
+			startWith(''),
+			debounceTime(300), // Optional: debounce input changes to avoid excessive requests
+			distinctUntilChanged(), // Optional: ensure distinct values before making requests
+			switchMap(value => this.space.getList(10, 0, value || '').pipe(
+				map(res => res.results)
+			))
+		);
+
+		this.controlAutocomplete.valueChanges.pipe(
+			debounceTime(500),
+			distinctUntilChanged(),
+		).subscribe(selected => {
+			if (!selected) {
+				this.selectedSpace = null;
+			}
+		});
+	}
+
+	onSelectOption(option: MatOption) {
+		if (option) {
+			this.selectedSpace = {
+				id: option.value.id,
+				name: option.value.name,
+				key: option.value.key,
+				requestsAllowed: option.value.requestAllowed,
+				createdAt: option.value.createdAt
+			}
+		}
+	}
+
+	sendRequestToJoin() {
+		const navigationExtras: NavigationExtras = {state: {spaceName: this.selectedSpace.name}};
+		this.router.navigate(['space/add/request'], navigationExtras);
+	}
+
+	private _transliterate(word: string): string {
+		return word
+			.split('')
+			.map(char => this.translitFromRuToEn[char] || char)
+			.join('');
+	}
+
+	onNameChange(name: string) {
+		if (!this.keyWasChanged) {
+			let newKey = name.toLowerCase();
+			newKey = this._transliterate(newKey);
+			newKey = newKey.replace(/[^a-z0-9_]/g, '');
+			this.createForm.patchValue({
+				key: newKey
+			});
+		}
+	}
+
+	onKeyChange(key: string) {
+		if (!this.keyWasChanged){
+			this.keyWasChanged = true;
+		}
```

Comment:
Само условие можно убрать для краткости, потому что перезапись keyWasChanged ни к чему плохому не приведёт.
```suggestion
		this.keyWasChanged = true;
```

Если считаешь что правильнее всё таки с условием, то можно убрать фигурные скобки, т.к. во всей функции есть только одно это условие и отсутствие фигурных скобок визуально ничего не ухудшит.
```suggestion
		if (!this.keyWasChanged)
			this.keyWasChanged = true;
```

Если фигурные скобки всётаки нужны, то нужно добавить пробел после круглых скобок.
```suggestion
		if (!this.keyWasChanged) {
			this.keyWasChanged = true;
		}
```

Reply:
Убрал условие.